crates/burn-train/src/metric/precision.rs

```diff
@@ -0,0 +1,130 @@
+    pub fn with_threshold(mut self, threshold: f64) -> Self {
+        self.threshold = Some(threshold);
+        self.top_k = None;
+        self
+    }
+
+    /// Sets the class average.
+    pub fn with_class_average(mut self, class_average: ClassAverageType) -> Self {
+        self.class_average = class_average;
+        self
+    }
+
+    /// Sets the top k.
+    pub fn with_top_k(mut self, top_k: usize) -> Self {
+        self.top_k = Some(top_k);
+        self.threshold = None;
+        self
+    }
+}
+
+impl<B: Backend> Default for PrecisionMetric<B> {
+    /// Creates a new metric instance with default values.
+    fn default() -> Self {
+        Self {
+            state: NumericMetricState::default(),
+            _b: PhantomData,
+            threshold: Some(0.5),
+            class_average: ClassAverageType::Micro,
+            top_k: None,
+        }
+    }
+}
+
+impl<B: Backend> Metric for PrecisionMetric<B> {
+    const NAME: &'static str = "Precision";
+    type Input = ClassificationInput<B>;
+    fn update(
+        &mut self,
+        input: &ClassificationInput<B>,
+        _metadata: &MetricMetadata,
+    ) -> MetricEntry {
+        let (predictions, targets) = input.clone().into();
+        let [sample_size, _] = input.predictions.dims();
+        let metric = ConfusionStats::new(
+            predictions,
+            targets,
+            self.threshold,
+            self.top_k,
+            self.class_average,
+        )
+        .precision();
+
+        self.state.update(
+            100.0 * metric,
+            sample_size,
+            FormatOptions::new(Self::NAME).unit("%").precision(2),
+        )
+    }
+
+    fn clear(&mut self) {
+        self.state.reset()
+    }
+}
+
+impl<B: Backend> Numeric for PrecisionMetric<B> {
+    fn value(&self) -> f64 {
+        self.state.value()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        ClassAverageType::{self, *},
+        Metric, MetricMetadata, Numeric, PrecisionMetric,
+    };
+    use crate::tests::{
+        dummy_classification_input,
+        ClassificationType::{self, *},
+        THRESHOLD,
+    };
+    use crate::TestBackend;
+    use burn_core::tensor::TensorData;
+    use rstest::rstest;
+
+    #[rstest]
+    #[case::binary_micro(Binary, Micro, 0.5)]
+    #[case::binary_macro(Binary, Macro, 0.5)]
+    #[case::multiclass_micro(Multiclass, Micro, 3.0/5.0)]
+    #[case::multiclass_macro(Multiclass, Macro, (0.5 + 0.5 + 1.0)/3.0)]
+    #[case::multilabel_micro(Multilabel, Micro, 5.0/8.0)]
+    #[case::multilabel_macro(Multilabel, Macro, (2.0/3.0 + 2.0/3.0 + 0.5)/3.0)]
+    fn test_precision(
+        #[case] class_type: ClassificationType,
+        #[case] avg_type: ClassAverageType,
+        #[case] expected: f64,
+    ) {
+        let input = dummy_classification_input(&class_type);
+        let mut metric = PrecisionMetric::<TestBackend>::default();
+        metric = match class_type {
+            Multiclass => metric.with_top_k(1),
```

Comment:
We should probably add a test for `top_k` > 1 to make sure everything is covered 🙂

Reply:
you're right, added